test/helpers/kubectl.go

```diff
@@ -1130,3 +1130,84 @@ func (kub *Kubectl) GetCiliumPodOnNode(namespace string, node string) (string, e
 
 	return res.Output().String(), nil
 }
+
+// CiliumPreFlightCheck specify that it checks that various subsystems within
+// Cilium are in a good state. If one of the multiple preflight fails it'll
+// return an error.
+func (kub *Kubectl) CiliumPreFlightCheck() error {
+	err := kub.CiliumControllersPreFlightCheck()
+	if err != nil {
+		return err
+	}
+	err = kub.CiliumHealthPreFlightCheck()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// CiliumControllersPreFlightCheck validates that all controllers are not
+// failing. If any of the controllers fails will return an error.
+func (kub *Kubectl) CiliumControllersPreFlightCheck() error {
+	var controllersFilter = `{range .controllers[*]}{.name}{"="}{.status.consecutive-failure-count}{"\n"}{end}`
+	ciliumPods, err := kub.GetCiliumPods(KubeSystemNamespace)
+	if err != nil {
+		return err
+	}
+	for _, pod := range ciliumPods {
+		status := kub.CiliumExec(pod, fmt.Sprintf(
+			"cilium status --all-controllers -o jsonpath='%s'", controllersFilter))
+		if !status.WasSuccessful() {
+			return fmt.Errorf("Cannot get status on %q: %s", pod, status.OutputPrettyPrint())
+		}
+		for controller, status := range status.KVOutput() {
+			if status != "0" {
+				return fmt.Errorf("Controller %s is failing in cilium", controller)
+			}
+		}
+	}
+	return nil
+}
+
+// CiliumHealthPreFlightCheck checks that the health status is working
+// correctly and the number of nodes does not mistmatch with the running pods.
+// It return an error if health mark a node as failed.
+func (kub *Kubectl) CiliumHealthPreFlightCheck() error {
+	var nodesFilter = `{.nodes[*].name}`
+	var statusFilter = `{range .nodes[*]}{.name}{"="}{.host.primary-address.http.status}{"\n"}{end}`
+
+	ciliumPods, err := kub.GetCiliumPods(KubeSystemNamespace)
+	if err != nil {
+		return err
+	}
+	for _, pod := range ciliumPods {
+		status := kub.CiliumExec(pod, "cilium-health status -o json --probe")
+		if !status.WasSuccessful() {
+			return fmt.Errorf("Cannot get status on %q: %s", pod, status.OutputPrettyPrint())
```

Comment:
When will this error trigger? When we failed to get status or when the cluster health is unhealthy? We need clear error messages that indicate either. We have lots of failures in the form of `Failed to get pods` when in fact one of the conditions is not being met and it's unclear which one. The test error should say:

```
Error: cluster connectivity is unhealthy
<< cilium-health status output >>
```